python/cudf/dataframe/categorical.py

```diff
@@ -256,33 +256,43 @@ def fillna(self, fill_value, inplace=False):
         """
         Fill null values with *fill_value*
         """
-
-        result = self.copy()
-
         if not self.has_null_mask:
-            return result
+            return self
 
         fill_is_scalar = np.isscalar(fill_value)
 
-        if fill_is_scalar and fill_value == self.default_na_value():
-            fill_value_col = columnops.as_column(
-                [fill_value], nan_as_null=False, dtype=self.data.dtype)
+        if fill_is_scalar:
+            if fill_value == self.default_na_value():
+                fill_value = self.data.dtype.type(fill_value)
+            else:
+                try:
+                    fill_value = self._encode(
+                        pd.Categorical(
+                            fill_value, categories=self.cat().categories
+                        )
+                    )
+                    fill_value = self.data.dtype.type(fill_value)
+                except (ValueError) as err:
+                    err_msg = "fill value must be in categories"
+                    raise ValueError(err_msg) from err
         else:
-            if fill_is_scalar:
-                if fill_value != self.default_na_value():
-                    if (fill_value not in self.cat().categories):
-                        raise ValueError("fill value must be in categories")
-                fill_value = pd.Categorical(fill_value,
-                                            categories=self.cat().categories)
-
-            fill_value_col = columnops.as_column(
-                fill_value, nan_as_null=False)
-
+            fill_value = columnops.as_column(fill_value, nan_as_null=False)
```

Comment:
If someone passes a Pandas / Numpy Series with `NaN`s here you don't think they'd want them converted to nulls?